pkg/urcrypt/urcrypt/pedersen.c

```diff
@@ -0,0 +1,356 @@
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p2), p2_buf, 64, crv);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p3), p3_buf, 64, crv);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_import_from_aff_buf(&(cp->p4), p4_buf, 64, crv);
+  return ret;
+}
+
+int
+do_hash(urcrypt_pedersen_context *cxt, uint8_t *a, size_t a_len, uint8_t *b, size_t b_len, uint8_t out[32])
+{
+  int ret = 0;
+
+  urcrypt__reverse(a_len, a);
+  urcrypt__reverse(b_len, b);
+
+  nn alow, ahig, blow, bhig;
+  if ( 32 == a_len ) {
+    ret = nn_init_from_buf(&alow, a+1, 31);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    ret = nn_init_from_buf(&ahig, a, 1);
+    if ( 0 != ret ) {
+      return ret;
+    }
+  }
+  else {
+    ret = nn_init(&ahig, 1);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    ret = nn_zero(&ahig);
+    if ( 0 != ret ) {
+      return ret;
+    }
+
+    ret = nn_init_from_buf(&alow, a, a_len);
+    if ( 0 != ret ) {
+      return ret;
+    }
+  }
+
+  if ( 32 == b_len ) {
+    ret = nn_init_from_buf(&blow, b+1, 31);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    *b = *b & 0x0f;
+    ret = nn_init_from_buf(&bhig, b, 1);
+    if ( 0 != ret ) {
+      return ret;
+    }
+  }
+  else {
+    ret = nn_init(&bhig, 1);
+    if ( 0 != ret ) {
+      return ret;
+    }
+    ret = nn_zero(&bhig);
+    if ( 0 != ret ) {
+      return ret;
+    }
+
+    ret = nn_init_from_buf(&blow, b, b_len);
+    if ( 0 != ret ) {
+      return ret;
+    }
+  }
+
+  prj_pt p1_alow, p2_ahig, p3_blow, p4_bhig;
+  ret = prj_pt_mul(&p1_alow, &alow, &(cxt->cp->p1));
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_mul(&p2_ahig, &ahig, &(cxt->cp->p2));
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_mul(&p3_blow, &blow, &(cxt->cp->p3));
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = prj_pt_mul(&p4_bhig, &bhig, &(cxt->cp->p4));
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  prj_pt sum;
+  ret = prj_pt_add(&sum, &(cxt->cp->p0), &p1_alow);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p2_ahig);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p3_blow);
+  if ( 0 != ret ) {
+    return ret;
+  }
+  ret = prj_pt_add(&sum, &sum, &p4_bhig);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  prj_pt aff;
+  ret = prj_pt_unique(&aff, &sum);
+  if ( 0 != ret ) {
+    return ret;
+  }
+
+  ret = fp_export_to_buf(out, 32, &(aff.X));
+  urcrypt__reverse(32, out);
+  return ret;
+}
+
+static int
+do_fold_hash(urcrypt_pedersen_context *cxt, uint8_t *dat_x, size_t len_x, uint8_t *out) {
+  int ret = 0;
+  int pos = 0, ext_pos = 0, ext_len = (len_x / 32);
+  uint8_t *hed = NULL, *acc = NULL, *ext = NULL;
+  while ( pos < len_x ) {
+    hed = (dat_x + pos);
+    if ( (len_x - pos) >= 32 && (hed[31] & 0xf0) != 0) {
+        if ( ext == NULL ) {
+          ext = calloc(1, ext_len);
```

Comment:
Do we need to conditionally allocate this inside the loop? Seems since we know its length, we're going to need to allocate it, no? I'm also a little confused by the calloc instead of malloc, and the arguments seem to be in the wrong order (not that that makes a difference).